folding-schemes/src/folding/hypernova/circuits.rs

```diff
@@ -345,24 +430,320 @@ fn compute_c_gadget<F: PrimeField>(
+    pub poseidon_config: PoseidonConfig<CF1<C1>>,
+    pub ccs: CCS<C1::ScalarField>, // CCS of the AugmentedFCircuit
+    pub i: Option<CF1<C1>>,
+    pub i_usize: Option<usize>,
+    pub z_0: Option<Vec<C1::ScalarField>>,
+    pub z_i: Option<Vec<C1::ScalarField>>,
+    pub external_inputs: Option<Vec<C1::ScalarField>>,
+    pub u_i_C: Option<C1>, // u_i.C
+    pub U_i: Option<LCCCS<C1>>,
+    pub U_i1_C: Option<C1>, // U_{i+1}.C
+    pub F: FC,              // F circuit
+    pub x: Option<CF1<C1>>, // public input (u_{i+1}.x[0])
+    pub nimfs_proof: Option<NIMFSProof<C1>>,
+}
+
+impl<C1, C2, GC2, FC> AugmentedFCircuit<C1, C2, GC2, FC>
+where
+    C1: CurveGroup,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+    FC: FCircuit<CF1<C1>>,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C2 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+    <C2 as Group>::ScalarField: Absorb,
+    C1: CurveGroup<BaseField = C2::ScalarField, ScalarField = C2::BaseField>,
+    for<'a> &'a GC2: GroupOpsBounds<'a, C2, GC2>,
+{
+    pub fn empty(
+        poseidon_config: &PoseidonConfig<CF1<C1>>,
+        F_circuit: FC,
+        ccs: Option<CCS<C1::ScalarField>>,
+    ) -> Result<Self, Error> {
+        let initial_ccs = CCS {
+            // m, n, s, s_prime and M will be overwritten by the `upper_bound_ccs' method
+            m: 0,
+            n: 0,
+            l: 2, // io_len
+            s: 1,
+            s_prime: 1,
+            t: 3, // note: this is only supports R1CS for the moment
+            q: 2,
+            d: 2,
+            S: vec![vec![0, 1], vec![2]],
+            c: vec![C1::ScalarField::one(), C1::ScalarField::one().neg()],
+            M: vec![],
+        };
+        let mut augmented_f_circuit = Self {
+            _c2: PhantomData,
+            _gc2: PhantomData,
+            poseidon_config: poseidon_config.clone(),
+            ccs: initial_ccs.clone(),
+            i: None,
+            i_usize: None,
+            z_0: None,
+            z_i: None,
+            external_inputs: None,
+            u_i_C: None,
+            U_i: None,
+            U_i1_C: None,
+            F: F_circuit,
+            x: None,
+            nimfs_proof: None,
+        };
+        if ccs.is_some() {
+            augmented_f_circuit.ccs = ccs.unwrap();
+        } else {
+            augmented_f_circuit.ccs = augmented_f_circuit.upper_bound_ccs()?;
+        }
+        Ok(augmented_f_circuit)
+    }
+
+    /// This method computes the CCS parameters. This is used because there is a circular
+    /// dependency between the AugmentedFCircuit CCS and the CCS parameters m & n & s & s'.
+    /// For a stable FCircuit circuit, the CCS parameters can be computed in advance and can be
+    /// feed in as parameter for the AugmentedFCircuit::empty method to avoid computing them there.
+    pub fn upper_bound_ccs(&self) -> Result<CCS<C1::ScalarField>, Error> {
```

Comment:
Sorry but I'm not sure on what is this exactly doing. 
Why do we need to iter `let n_iters = 3;` always?

Reply:
This is because in HyperNova there is a circular dependency (there is [this comment in the code](https://github.com/privacy-scaling-explorations/sonobe/blob/2e7f7893b9f9fe844cb035e09fcd5010ce84c09a/folding-schemes/src/folding/hypernova/circuits.rs#L516) explaining it a bit, but I will now update to extend it with the current comment that I'm writing here to provide more context):

the HyperNova's AugmentedFCircuit size depends on the number of rounds of the SumCheck determined by the $s$ parameter, which depends on the size of the AugmentedFCircuit size.
Since for each FCircuit the AugmentedFCircuit will be different, its size needs to be computed each time, and a way was by combining iterating the constraint generation then generating the ccs out of it and using it to compute the constraint generation again, a total of 3 times. (with n_iters=2 it works for example when the sumcheck constraints don't make the difference between a power of two and the next one, but when for example the number of constraints is just a bit less of a power of two, depending on the number of constraints, the circular dependency will kick in and will make the number of constraints grow a bit, but not being reflected yet in the CCS parameters. So with n_iters=3 it gets enough iterations to ensure that the number of constraints (which depend on the correct CCS parameters) will be reflected in the CCS parameters.

That's why in the `AugmentedFCircuit::empty` method the `CCS` can be feed in as input: [github.com/privacy-scaling-explorations/sonobe/folding-schemes/src/folding/hypernova/circuits.rs#L475](https://github.com/privacy-scaling-explorations/sonobe/blob/2e7f7893b9f9fe844cb035e09fcd5010ce84c09a/folding-schemes/src/folding/hypernova/circuits.rs#L475) , so that for stabilized FCircuit circuits, the CCS can be computed beforehand and passed to this method to avoid computing it each time.

The idea is that this method is only used to prepare the CCS parameters before real-world usage, and for a stable circuit this is not used since the CCS parameters are stable and passed as input.

(will add this comment as comment in the code for future readers to provide context)

Reply:
One note on this: I think that anyways this will be affected once CycleFold is added soon, I will have it mind for when working on it.